### PR TITLE
Updates on new-wpinterpolant

### PR DESCRIPTION
**Pull request by ArpitaDutta**
The base is the master branch (after merging) and recent commits are cherrypicked from new-wpinterpolant branch.

### CODE REVIEW

File: lib/Core/TxTree.cpp [hunk 1]
```diff
@@ -2434,6 +2434,10 @@ void TxTree::remove(ExecutionState *state, TimingSolver *solver, bool dumping) {
         if (!WPExpr.isNull()) {
           entry = node->wp->updateSubsumptionTableEntry(entry);
         }
+        if(node->assertionFail){
+					ref<Expr> resetWPExpr = ConstantExpr::alloc(0, Expr::Bool);
+					entry->setWPInterpolant(resetWPExpr);
+				}
```
Comment (rasoolmaghareh) [inline lib/Core/TxTree.cpp:2440]: Fix the code style

File: lib/Core/TxTree.cpp [hunk 2]
```diff
@@ -2870,50 +2874,49 @@ ref<Expr> TxTreeNode::instantiateWPatSubsumption(ref<Expr> wpInterpolant,
...
-        }
-      } else { // Date: 20/04/2022
-               //  // Loading values from Global Variables
-        ref<Expr> dummy;
-        //        llvm::outs()<<"\nChecking the Global list for the WPVar:
-        //        ["<<wpInterpolant<<"]\n";
-        //        llvm::outs()<<"cheking WP var address\n";
-        //        llvm::outs()<<WPVar1->address<<"\n";
-        //        wpInterpolant.get()->dump();
-        //        llvm::outs()<<wpInterpolant.get();
```
Comment (rasoolmaghareh) [inline lib/Core/TxTree.cpp:2897]: remove the unwanted comments

File: lib/Core/TxTree.cpp [hunk 3]
```diff
@@ -2870,50 +2874,49 @@ ref<Expr> TxTreeNode::instantiateWPatSubsumption(ref<Expr> wpInterpolant,
...
+           }
+         }
+         else{ // Date: 20/04/2022
+        	//  // Loading values from Global Variables
+                  	  ref<Expr> dummy;
+          //        llvm::outs()<<"\nChecking the Global list for the WPVar: ["<<wpInterpolant<<"]\n";
+          //        llvm::outs()<<"cheking WP var address\n";
+          //        llvm::outs()<<WPVar1->address<<"\n";
+          //        wpInterpolant.get()->dump();
+          //        llvm::outs()<<wpInterpolant.get();
```
Comment (rasoolmaghareh) [inline lib/Core/TxTree.cpp:2899]: remove the unwanted comments

File: lib/Core/TxTree.cpp [hunk 4]
```diff
@@ -2870,50 +2874,49 @@ ref<Expr> TxTreeNode::instantiateWPatSubsumption(ref<Expr> wpInterpolant,
...
+                        	 ref<Expr> address=ConstantExpr::create(mo->address, Expr::Int64);
+                        	 ref<Expr> offset = mo->getOffsetExpr(address);
+                	 		     ref<Expr> result = oj->read(offset, mo->size*8);
+                	 		     return result;
+                	 	 }
+                 }
+                 }
+        klee_warning("TxTreeNode::instantiateWPatSubsumption: Instantiation failed!");
+              return dummy;
+        //llvm::outs()<<globalAddresses;
```
Comment (rasoolmaghareh) [inline lib/Core/TxTree.cpp:2917]: remove the comment

File: lib/Core/TxTree.cpp [hunk 5]
```diff
@@ -3020,123 +3024,296 @@ ref<Expr> TxTreeNode::instantiateWPatSubsumption(ref<Expr> wpInterpolant,
...
+            // 1;
+            //						if (newindex <
+            // totalIndex)
+            //{
+            //							//correctindex =
+            // true;
+            //							offset =
+            // ConstantExpr::create(newindex,
+            //									Expr::Int64);
+            //						}
```
Comment (rasoolmaghareh) [inline lib/Core/TxTree.cpp:3297]: please remove comments & run clang-format

File: lib/Core/Z3Simplification.cpp [hunk 6]
```diff
@@ -287,17 +296,63 @@ bool Z3Simplification::txExpr2z3Expr(z3::expr &z3e, z3::context &c,
...
+			  // std::string strI = std::to_string(Array_index);
+			  // std::string tag = "selExpr_";
+			  // std::string name2insert=tag+Array_name+"_"+strI;
+        // //std::cout<<"name2insert:"<<name2insert<<"\n";
+			  // z3e=c.int_const(name2insert.c_str());
+			  // emap.insert(std::pair<std::string, ref<Expr> >(name2insert, txe));
+			  // return true;
+      	//   }
+       	//   }
+       	//   klee_warning("Cannot convert tx::sel to z3 expr");
```
Comment (rasoolmaghareh) [inline lib/Core/Z3Simplification.cpp:338]: remove comments

File: lib/Core/Z3Simplification.cpp [hunk 7]
```diff
@@ -287,17 +296,63 @@ bool Z3Simplification::txExpr2z3Expr(z3::expr &z3e, z3::context &c,
...
+       	  return false;
+         }
+         case Expr::SExt: {
+          //  z3::expr t1 = c.bool_val(false);
+          //  bool r1 = txExpr2z3Expr(t1, c, txe->getKid(0), emap);
+          //  if (r1) {
+          //    z3e = sext(t1, txe->getWidth());
+          //    return true;
+          //  }
+          //  klee_warning("Cannot convert tx::sel to z3 expr");
```
Comment (rasoolmaghareh) [inline lib/Core/Z3Simplification.cpp:348]: remove comments

File: lib/Core/Z3Simplification.cpp [hunk 8]
```diff
@@ -368,6 +423,7 @@ Z3Simplification::z3Expr2TxExpr(z3::expr e,
       if (!val0.compare(l) || !val0.compare(r)) {
         ref<Expr> val1 = ConstantExpr::create(1, Expr::Int32);
         ref<Expr> val2 = SDivExpr::create(val0, val1);
+        // val2->dump();
```
Comment (rasoolmaghareh) [inline lib/Core/Z3Simplification.cpp:426]: remove comment

**Review by rasoolmaghareh — COMMENTED** [review]
Please also post your test results on this branch compared to master branch.

File: lib/Core/Z3Simplification.cpp (hunk 8 above)
Comment (ArpitaDutta) [reply lib/Core/Z3Simplification.cpp:426]: Sure. Working on it.

**Comment by ArpitaDutta** [comment]
> Please also post your test results on this branch compared to master branch.

The comparison result with master-branch wp is [here](https://docs.google.com/spreadsheets/d/1L4ae6hbE8JN0Y55MGhZ7-LdUWEkdCr3ajFiwspkXvSM/edit?usp=sharing).

**Comment by ArpitaDutta** [comment]
Updated the [file](https://docs.google.com/spreadsheets/d/1L4ae6hbE8JN0Y55MGhZ7-LdUWEkdCr3ajFiwspkXvSM/edit?usp=sharing) with master branch deletion interpolation result.

**Comment by ArpitaDutta** [comment]
There are differences in the number of covered blocks and icmps for programs that have hit the timeout.

**Comment by rasoolmaghareh** [comment]
> There are differences in the number of covered blocks and icmps for programs that have hit the timeout.

The results are good. All cases that have differences are reaching timeout, or have a meaningful explanation on the difference. 

I think we can proceed with merging this MR after addressing the comments on formatting and squashing the commits.

**Comment by ArpitaDutta** [comment]
I have made this commit [c75c384](https://github.com/tracer-x/TracerX/pull/391/commits/c75c38472af993e91a17740bd885e5413d7963ba) today to remove the comments and formatting the code using clang.

**Comment by rasoolmaghareh** [comment]
> I have made this commit [c75c384](https://github.com/tracer-x/TracerX/pull/391/commits/c75c38472af993e91a17740bd885e5413d7963ba) today to remove the comments and formatting the code using clang.

clang formatting seems not have been done correctly in `lib/Core/TxTree.cpp`

**Comment by ArpitaDutta** [comment]
I have disabled the assertion violation-related commit ([1d5f8d3](https://github.com/tracer-x/TracerX/pull/391/commits/1d5f8d3451aefa3371c105e565da958e0c677fde)).
Now trying to correctly format the lib/Core/TxTree.cpp file.

**Comment by ArpitaDutta** [comment]
Tried to reformat the lib/Core/TxTree.cpp file.
Please let me know if there is any change required.

**Review by rasoolmaghareh — APPROVED** [review]

**Pull request merged**